Win client

domderen opened the pull request:
Some details can be found at: https://github.com/kitematic/kitematic/issues/312

DONE

* I recreated the deps and reset scripts for windows to install and remove external dependencies. They are written in powershell, coz my CMD skills really suck, and I am not sure if some steps could be done in CMD without at least invoking the PS commands. I was able to loose Chocolatey dependency though, which required administrator rights, so right now the whole setup can be done without admin rights.
* I have removed the pre and post install scripts from package.json as this is really discuraged by the NPM docs, and instead moved them to gulpfile. This also allowed me to port those scripts for Windows platform.
* I've updated the SetupStore and SetupUtil to prepare the application for windows, including the downloading and installation of VirtualBox. As mentioned before, it is done without Chocolatey, so it still might be a little shaky.
* Updated the list of directories used by the app, to proper Windows paths, so the docker and docker-machine binaries are placed on the user's directory. This might not be the best place for them, but it is a start.
* I noticed that docker-machine's boot2docker image, does not automatically share Users directory with the VM, so I created a workaround and invoke a VirtualBox command to do it manually during the VM initialization step in the SetupStore. This is a strange one though coz the regular boot2docker does that automatically. Maybe this is an issue to be raised in the docker-machine repo.
* I've rewritten a bit parts of code responsible for opening URLs or drive paths, and added an extra dependency node package called "open" which supports different platforms out of the box, so opening links and paths should now work on all platforms.
* Ability to open a CMD window with Docker env variables set up, and docker on the path.
* Tested the application setup process on WIndows 8.1 and it works as expected.

So basically now it should be possible to clone the repo, install npm deps, download app deps using the PS script, start the application in development mode, have VirtualBox installed and VM initialized, and next it should be possible to create a new container with files shared from the windows host.

AWAITS
* File sharing and selecting directories in the app is flaky on windows, it still needs some magic to work properly. It works with the directory that is attached on container creation, but it is not possible yet, to change the default dir. Also the naming for the local directory in the container's settings should be changed, coz it is called "Mac Folder", so it is not really friendly for different platforms :)
* Application deploy in gulp file, is not ready for deploying windows version. I wanted to leave it as the last piece.
* Testing if any of my changes don't break any of the Mac functionality. Unfortunately I don't have access to Mac, so I am only testing the app on Windows, so I don't know if I didn't break anything.
* Probably some graphical changes to the app window itself to make it more Windows like :)
* Adding support for ssh terminal in ContainerSubHeader.

domderen commented:
Hmm, it looks like the test should also be fixed, to properly read VBox Version and VBox Filename from package.json.

jeffdm commented:
Awesome! This is amazing progress towards Windows!  No rush to fix the tests.

I'll clone this and get it up later today. Following that I'll run some of our test cases for the setup. So far, so good though! Wow!

Elesant commented:
Wow. This is amazing @domderen ! I'll play with this a bit on Windows

ahmetalpbalkan commented:
Crazy!! cc:@tianon for b2d related issues

domderen commented:
I've added support for opening up a terminal window on windows. It uses the CMD for that purpose and sets up the required environment variables.

jeffdm commented:
@domderen thanks for fixing the tests. Awesome :). Also, I really appreciate some of the overall code improvements (proper use of npm scripts, etc.). This is fantastic.

jeffdm commented:
Great! Wow!

domderen commented:
I am glad to hear that :) I was wondering if you have any designers at the team, who could have a look at changing the app window for Windows systems to be more consistent? Or maybe Atom-Shell provides something meaningful?

jeffdm commented:
Sean (@Elesant) designs Kitematic. He can whip up an awesome Windows header design for Kitematic

Elesant commented:
Already working on it :)

domderen commented:
Cool :)

domderen commented:
Hey,

I've refactored the SetupUtil and especially installation step on windows, as it didn't work correctly. Now the whole setup process should work from beginning to end as expected on windows. The complete list of changes made:

* Added Resources module which provides paths to all files stored in the resources directory,
* Added docker and docker-machine binary paths to Util module, to have a single place for calling those paths,
* Added binaries directory path, to Util module, to have a single place for calling this path,
* Added another exec command, which utilizes "child_process" module, and it's "exec" command, as the "exec" used right now is deprecated,
* Refactored the Installation step in the SetupStore, and connected with that SetupUtil commands to perform copying binaries from node instead of bash commands. That way application is more cross-platform friendly,
* Added new dependencies in form of fs-extra, fs-promise and any-promise, to allow yielding of fs commands in SetupUtil.
* The setup process now works correctly on windows from the beginning to end.

jeffdm commented:
@domderen ! This is amazing! I'm testing this out right now!

jeffdm commented:
In the meantime, I'm reading through the code and just wanted to say wow. This doesn't just add windows but it also reduces the tech debt and introduces some better practices and code design to the project.

Thanks @domderen !

jeffdm commented:
Up and running! This is amazing :smiley: 

![screen shot 2015-03-29 at 3 38 06 pm](https://cloud.githubusercontent.com/assets/251292/6887325/c86b504a-d629-11e4-94e0-a1b3df645f40.png)

DieterReuter commented:
WOW, I'm really impressed! 
Great work @domderen!

ahmetalpbalkan commented:
@domderen that's crazy!

kmouss commented:
Awesome!

StefanScherer commented on package.json:
```diff
@@ -45,19 +43,25 @@
   "atom-shell-version": "0.21.3",
   "virtualbox-version": "4.3.24",
   "virtualbox-filename": "VirtualBox-4.3.24.pkg",
```
Great, you're using VirtualBox 4.3.26 on Windows. In another PR we should update from 4.3.24 to 4.3.26 on Mac as well.

jeffdm commented on appveyor.yml:
```diff
@@ -0,0 +1,21 @@
+# Test against this version of Node.js
```
Since jest just tests the javascript module we don't need a windows environment just _yet_ (everything is mocked).

Won't hurt to have this file though. Awesome!!

jeffdm commented on src/ContainerDetailsSubheader.react.js:
```diff
@@ -89,7 +91,7 @@ var ContainerDetailsSubheader = React.createClass({
         from: 'header'
       });
-      exec(['open', this.state.ports[this.state.defaultPort].url], function (err) {
+      util.openPathOrUrl(this.state.ports[this.state.defaultPort].url, function (err) {
```
Wondering if we could use: https://github.com/atom/atom-shell/blob/master/docs/api/shell.md

jeffdm commented on src/SetupStore.js:
```diff
@@ -70,9 +68,33 @@ var _steps = [{
         yield machine.rm();
         yield machine.create();
+        if(util.isWindows()) {
+          var home = util.home();
```
Any way we can move this to setuputil?

jeffdm commented:
Added some minor comments but this looks great. Let's get it merged as a checkpoint!

One more thing before merging, would it be possible to bundle these as one commit and [sign it](https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work)?

It's to certify that you wrote the changes or otherwise have the right to pass it on as an open-source patch.

Thanks @domderen ! Amazing.

domderen commented:
@jeffdm I am a bit new to the concept of signing the commits, so you would have to guide me through it. Should I revert current commits, and push them all as one with a '-s' parameter in git?

ahmetalpbalkan commented:
@domderen you need to squash all your commits into one and run "git commit --amend -s --no-edit" (which will sign the commit) and "git push -f" to your win_client branch.

domderen commented:
@ahmetalpbalkan This squash all commits into one is a part I have never done before, I'd appreciate some help here.

domderen commented:
Hey, I've done something, though I am not sure that is what you were looking for.

ahmetalpbalkan commented:
@domderen hmm you need to fetch kitematic's remote && rebase kitematic's master while you're on your branch and solve resolve conflicts if any and git push -f again. It appears like you dodnt start with a clean branch off of up to date master in this PR.

domderen commented:
Yeah but I did manage to push a commit here with all the changes and a sign off here. Isn't that enough?

jeffdm commented:
LGTM

Awesome start at Windows! Ready to merge. This is incredible @domderen

domderen commented:
Now should be fine :)

jeffdm commented:
Merging! Wooo :tada: :fireworks:

ahmetalpbalkan commented:
@jeffdm Are you folks planning to release some pre-release windows executable? I'd love to try out and give feedback.

jeffdm commented:
Good idea! Will upload one to GitHub.

rrarunan commented:
:tada: Awesome!

laurentkempe commented:
@jeffdm Do you have any link to share for the download?

Oliboy50 commented:
@jeffdm The new [docker documentation](http://docs.docker.com/kitematic/) is out and talks about kitematic on Windows, but I still can't find any link for downloading this (awesome, i guess) app for Windows. 
The link from the doc (https://kitematic.com/download/) gives me the zipped Mac app

ahmetalpbalkan commented:
@Oliboy50  It's in alpha and you should sign up: http://blog.docker.com/2015/06/kitematic-windows-alpha/

Oliboy50 commented:
@ahmetalpbalkan I already signed up for this survey but nothing came out at this time... Maybe I didn't enter my email address... Well I did it again just now with my email address, hope I'll receive a mail with a link or something. Thanks

Oliboy50 commented:
Hey the Kitematic Windows release is now officially available for download : https://kitematic.com/download/ 

Thank you guys :)
